src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -2937,6 +2941,27 @@ void ObjectSynchronizer::chk_per_thread_in_use_list_and_count(JavaThread *jt,
   }
 }
 
+// Do the final audit and print of ObjectMonitor stats; must be done
+// by the VMThread (at VM exit time).
+void ObjectSynchronizer::do_final_audit_and_print_stats() {
+  assert(Thread::current()->is_VM_thread(), "sanity check");
+
+  if (is_final_audit()) {  // Only do the audit once.
```

Comment:
It seems impossible that you could ever call this twice. A simple file static flag for a debug build would be simpler if you just wanted to assert that.

Reply:
The final_audit flag seems overkill. If the audit code needs to know it is the final audit that could be passed as a parameter.
The make-sure-this-is-only-called-once logic gives the false impression that it is entirely possible for this to be called more than once when that is not the case. We can only reach this code after the termination safepoint has been reached. A simple assert would be clearer IMO, but I don't insist on any changes here.